src/subscription/base-subscription.ts

```diff
@@ -0,0 +1,316 @@
+import { TokenProvider } from '../token-provider';
+import { RetryStrategy } from '../retry-strategy/retry-strategy';
+import { Logger } from '../logger';
+import { XhrReadyState, NetworkError, ErrorResponse, Headers, responseHeadersObj } from "../base-client";
+
+export enum SubscriptionState {
+    UNOPENED = 0, // haven't called xhr.send()
+    OPENING,      // called xhr.send(); not yet received response headers
+    OPEN,         // received response headers; called onOpen(); expecting message
+    ENDING,       // received EOS message; response not yet finished
+    ENDED         // called onEnd() or onError(err)
+}
+
+export interface SubscriptionEvent {
+    eventId: string;
+    headers: Headers;
+    body: any;
+}
+
+export function createSubscriptionConstructor(
+    retryStrategy: RetryStrategy, 
+    headers: Headers, 
+    xhrSource: () => XMLHttpRequest): 
+    (error: any, lastEventId?: string) => BaseSubscriptionConstruction {
+        return  (error: any, lastEventId?: string) => { 
+            return new BaseSubscriptionConstruction(retryStrategy, xhrSource, error, lastEventId); 
+        };   
+    }
+    
+    export class BaseSubscriptionConstruction {
+        private subscription: BaseSubscription;
+        private error: any;
+        private subscriptionCallback: (subscription: BaseSubscription) => void;
+        private errorCallback: (error: any) => void;
+        
+        constructor(
+            private retryStrategy: RetryStrategy,
```

Comment:
Why does base subscription constructor take a retry strategy? I thought it was going to be the most primitive subscription implementation, simply failing on error.